src/components/CytoscapeGraph/CytoscapeGraph.tsx

```diff
@@ -206,57 +198,66 @@ export class CytoscapeGraph extends React.Component<CytoscapeGraphProps, Cytosca
 
     this.trafficRenderer.stop();
 
+    const isTheGraphSelected = this.cy.$(':selected').length === 0;
+
     cy.startBatch();
 
+    // Destroy badges
+    // We must destroy all badges before updating the json, or else we will lose all the
+    // references to removed nodes
+    const cbBadge = new GraphBadge.CircuitBreakerBadge();
+    const rrBadge = new GraphBadge.RouteRuleBadge();
+    const rrGroupBadge = new GraphBadge.RouteRuleGroupBadge();
+    const msBadge = new GraphBadge.MissingSidecarsBadge();
+    cy.nodes().forEach(ele => {
+      cbBadge.destroyBadge(ele);
+      rrBadge.destroyBadge(ele);
+      rrGroupBadge.destroyBadge(ele);
+      msBadge.destroyBadge(ele);
+    });
+
     // update the entire set of nodes and edges to keep the graph up-to-date
     cy.json({ elements: this.props.elements });
 
-    // update the layout if it changed
-    if (this.newLayout) {
-      cy.layout(LayoutDictionary.getLayout(this.newLayout)).run();
-      this.newLayout = '';
+    // update the layout if needed
+    if (this.updateLayout) {
+      cy.layout(LayoutDictionary.getLayout(this.props.graphLayout)).run();
+      // Don't allow a large zoom if the graph has a few nodes (nodes would look too big).
+      if (cy.zoom() > 2.5) {
+        cy.zoom(2.5);
+        cy.center();
```

Comment:
Sometimes it doesn't seem like we center the graph outside of this use case.  Perhaps it is supposed to be centered by default, but it doesn't always seem that way to me when I change layouts.

Reply:
Maybe is related to #392

Reply:
We don't always want to force a center() - that would reset the pan thus jumping the graph when the user might not want to. If I drag the graph around, and I change the duration, or some other setting, I want the graph to remain there if possible (and this PR now does that!). I don't want the graph to jump to center when I just change the duration.

That said, if you change the LAYOUT, now you are asking Cy to redraw the entire graph. I don't think we can control much of that. IF we want to re-center the graph when changing the LAYOUT, I would be OK with that, since the entire graph is essentially going to change. This is also why I think its OK to reset the zoom, too, when changing the layout. The graph just changes too dramatically to expect the current zoom/pan levels to make sense with a new layout.

I would not worry about this for this PR. Let's get this in, as it is very nice as-is.

Reply:
+1, it was only a comment.  Let's get this merged as soon as @mtho11 approves.  But as I said above, I'm talking about on layout change, not on every update.